### PR TITLE
Implement Reward Based Ads

### PR DESCRIPTION
**Pull request by canbican**
I added support for reward based ads. Will be more than happy if you accept this contribution. Android part is coming soon.

### CODE REVIEW

File: ios/Classes/TiAdmobModule.m [hunk 1]
```diff
@@ -72,5 +72,6 @@ -(void)didReceiveMemoryWarning:(NSNotification*)notification
 MAKE_SYSTEM_PROP(GENDER_UNKNOWN, kGADGenderUnknown);
 MAKE_SYSTEM_PROP(AD_TYPE_BANNER, TiAdmobAdTypeBanner);
 MAKE_SYSTEM_PROP(AD_TYPE_INTERSTITIAL, TiAdmobAdTypeInterstitial)
+MAKE_SYSTEM_PROP(AD_TYPE_REWARDBASED, TiAdmobAdTypeRewardBased)
```
Comment (hansemannn) [inline ios/Classes/TiAdmobModule.m]: Use `AD_TYPE_REWARD_BASED` to match the code-style.

File: ios/Classes/TiAdmobView.m [hunk 2]
```diff
@@ -97,8 +97,11 @@ - (void)initialize
     
     if ([TiUtils intValue:adType def:TiAdmobAdTypeBanner] == TiAdmobAdTypeBanner) {
         [[self bannerView] loadRequest:[self request]];
-    } else {
+    } else if ([TiUtils intValue:adType def:TiAdmobAdTypeBanner] == TiAdmobAdTypeInterstitial) {
         [[self interstitial] loadRequest:[self request]];
+    } else {
+        [GADRewardBasedVideoAd sharedInstance].delegate = self;
+        [[GADRewardBasedVideoAd sharedInstance] loadRequest:[GADRequest request] withAdUnitID: [[self proxy] valueForKey:@"adUnitId"] userID: nil];
```
Comment (hansemannn) [inline ios/Classes/TiAdmobView.m]: Can you expose the `userID` as well, so the developer can pass both if desired?

File: ios/Classes/TiAdmobView.m [hunk 3]
```diff
@@ -224,6 +227,13 @@ - (void)showInterstitial
     }
 }
 
++ (void)showRewardBased
```
Comment (hansemannn) [inline ios/Classes/TiAdmobView.m]: Rename to `showRewardBasedAd`

File: ios/Classes/TiAdmobView.m [hunk 4]
```diff
@@ -353,4 +363,37 @@ - (void)interstitialWillLeaveApplication:(GADInterstitial *)ad
     [self.proxy fireEvent:@"willLeaveApplication" withObject:[TiAdmobView dictionaryFromInterstitial:ad]];
 }
 
+#pragma mark - RewardBased Delegate
+
+- (void)rewardBasedVideoAd:(GADRewardBasedVideoAd *)rewardBasedVideoAd
+   didRewardUserWithReward:(GADAdReward *)reward {
+    [self.proxy fireEvent:@"receiveReward" withObject:@{@"rewardType": reward.type , @"rewardAmount": reward.amount}];
```
Comment (hansemannn) [inline ios/Classes/TiAdmobView.m]: Did you test this? Does the decimal-number `reward.amount` get returned correctly to the frontend?

File: ios/Classes/TiAdmobView.m [hunk 5]
```diff
@@ -353,4 +363,37 @@ - (void)interstitialWillLeaveApplication:(GADInterstitial *)ad
...
+}
+
+- (void)rewardBasedVideoAdDidReceiveAd:(GADRewardBasedVideoAd *)rewardBasedVideoAd {
+    [self.proxy fireEvent:@"receiveAd" withObject:@{@"isReady": NUMBOOL([[GADRewardBasedVideoAd sharedInstance] isReady])}];
+}
+
+- (void)rewardBasedVideoAdDidOpen:(GADRewardBasedVideoAd *)rewardBasedVideoAd {
+    [self.proxy fireEvent:@"open"];
```
Comment (hansemannn) [inline ios/Classes/TiAdmobView.m]: We might want to add the type of add and / or ad-unit-id to the open-event to identify different types of ads. What do you think?

File: ios/Classes/TiAdmobView.m [hunk 6]
```diff
@@ -353,4 +363,37 @@ - (void)interstitialWillLeaveApplication:(GADInterstitial *)ad
...
+}
+
+- (void)rewardBasedVideoAdDidStartPlaying:(GADRewardBasedVideoAd *)rewardBasedVideoAd {
+    [self.proxy fireEvent:@"startPlaying"];
+}
+
+- (void)rewardBasedVideoAdDidClose:(GADRewardBasedVideoAd *)rewardBasedVideoAd {
+    [self.proxy fireEvent:@"closed"];
```
Comment (hansemannn) [inline ios/Classes/TiAdmobView.m]: Use `close` for general parity.

File: ios/Classes/TiAdmobViewProxy.h [hunk 7]
```diff
@@ -14,4 +14,6 @@
 
 - (void)receive:(id)unused;
 
+- (void)showRewardBased:(id)unused;
```
Comment (hansemannn) [inline ios/Classes/TiAdmobViewProxy.h]: Same here, rename the public API to `showRewardBasedAd` and adjust the docs.

File: ios/example/app.js [hunk 8]
```diff
@@ -78,6 +78,43 @@ btn.addEventListener('click', function() {
...
+    });
+    ad3.addEventListener("willLeaveApplication", function(e) {
+      Ti.API.debug("will leave application");
+    });
+    ad3.addEventListener("failedToLoad", function(e) {
+      Ti.API.debug("failed to load ad ",e);
+    });
+});
+
+win.add(btnra);
```
Comment (hansemannn) [inline ios/example/app.js]: I noticed you didn't pass any layout values to the button. Wouldn't it overlay with other buttons? Didn't do the FT so far.

File: ios/Classes/TiAdmobModule.m (hunk 1 above)
Comment (canbican) [inline ios/Classes/TiAdmobModule.m]: Changed in my next commit.

File: ios/Classes/TiAdmobView.m (hunk 2 above)
Comment (canbican) [inline ios/Classes/TiAdmobView.m]: Done.

File: ios/Classes/TiAdmobView.m (hunk 3 above)
Comment (canbican) [inline ios/Classes/TiAdmobView.m]: Done.

File: ios/Classes/TiAdmobView.m (hunk 4 above)
Comment (canbican) [inline ios/Classes/TiAdmobView.m]: It's not really decimal, admit accepts only integers, it seems. And yes it works.

File: ios/Classes/TiAdmobView.m (hunk 5 above)
Comment (canbican) [inline ios/Classes/TiAdmobView.m]: I believe in putting in different listeners :) And I'm not sure if it's conveniently possible -- underlying implementation is a singleton.

File: ios/Classes/TiAdmobView.m (hunk 6 above)
Comment (canbican) [inline ios/Classes/TiAdmobView.m]: Done.

File: ios/Classes/TiAdmobViewProxy.h (hunk 7 above)
Comment (canbican) [inline ios/Classes/TiAdmobViewProxy.h]: Done.

File: ios/example/app.js (hunk 8 above)
Comment (canbican) [inline ios/example/app.js]: I changed the window layout to vertical, I think it is sufficient.

File: ios/Classes/TiAdmobView.m (hunk 4 above)
Comment (canbican) [inline ios/Classes/TiAdmobView.m]: admit -> admob

**Comment by canbican** [comment]
Android version is due next week.

**Comment by canbican** [comment]
It turns out that the android version is way beyond the iOS one - too much gap to close. I took the liberty of implementing it as a separate module.

**Pull request closed without merging**